<!DOCTYPE html>
<html lang="pt-br">

<head>
  <title>Iniciando com o ORM Pony no Python III - Erros e Exceções</title>

  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1" >

  <meta name="description" content="Seguindo o trilha com o ORM Pony, neste terceiro texto, agora começaremos a gerar situações de erro e ver o que os objetos retornam ou que exceções são geradas.">
  <meta name="author" content="Paulo Henrique Rodrigues Pinheiro">

  <link rel="alternate" type="application/rss+xml" title="RSS Feed for Paulo Henrique Rodrigues Pinheiro - Programação para programadores!" href="https://paulohrpinheiro.xyz/feed.xml">

  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bulma@0.9.2/css/bulma.min.css">
  <script defer src="https://use.fontawesome.com/releases/v5.3.1/js/all.js"></script>

  <link rel=icon href="/favicon.ico" sizes="16x16" type="image/png">

  <link rel="stylesheet" href="https://paulohrpinheiro.xyz/assets/css/normalize.css">
</head>

<body>
  <section class="section">
    <div class="container">
      <h1 class="title is-1">
        Paulo Henrique Rodrigues Pinheiro
      </h1>
      <p class="subtitle is-3">
        Um blog sobre programação para programadores!
      </p>
    </div>
  </section>



     <div class="container">
        
		<nav class="level">
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/">HOME</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/c/">c</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/diversos/">diversos</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/go/">go</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/haskell/">haskell</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/javascript/">javascript</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/paremias/">paremias</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/perl/">perl</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/programadorbipolar/">programadorbipolar</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/prosaeverso/">prosaeverso</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/python/">python</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/ruby/">ruby</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/rust/">rust</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/server/">server</a>
			</p>
			<p class="level-item has-text-centered">
				<a class="link is-info content" href="https://paulohrpinheiro.xyz/texts/v/">v</a>
			</p>
		</nav>
	</div>

    <hr>


  <article class="container">
    <div class="content">
        <h2>Iniciando com o ORM Pony no Python III - Erros e Exceções</h2>

<h3>Seguindo o trilha com o ORM Pony, neste terceiro texto, agora começaremos a gerar situações de erro e ver o que os objetos retornam ou que exceções são geradas.</h3>

<p><img src="images/pony.png" alt="Logo do projeto Pony"></p>

<blockquote>
<p><a href="https://ponyorm.org/">https://ponyorm.org</a></p>
</blockquote>

<p>No <a href="https://paulohrpinheiro.xyz/texts/python/2021-04-28-iniciando-com-o-orm-pony-no-python.html">primeiro texto desta série</a>, demos os primeiros passos com o <a href="https://ponyorm.org">ORM Pony</a>, usando o <a href="https://sqlite.org/">SQLite</a> como banco de dados.</p>

<p>Já no <a href="https://paulohrpinheiro.xyz/texts/python/2021-05-11-iniciando-com-o-orm-pony-no-python-ii-banco-de-dados-com-docker.html">segundo texto desta coleção</a>, criamos um pequeno conjunto de dados a partir de uma requisição à <a href="https://api.github.com/">API do Github</a>, e usamos em três bancos distintos: <a href="https://www.mysql.com/">MySQL</a>, <a href="https://www.postgresql.org/">Postgres</a> e o bom e velho <a href="https://sqlite.org/">SQLite</a>, usando <a href="https://docs.docker.com/compose/">docker</a>.</p>

<p>Agora vamos fazer um monte de coisa errada para entender as situações de erro e estarmos preparados para programarmos essas situações.</p>

<h4>O Laboratório</h4>

<p>Partamos do laboratório construído no segundo texto acima referenciado. Por gosto pessoal, escolho o <em>Postgres</em> para conectar, mas os exemplos devem funcionar com qualquer um dos bancos, basta trocar a conexão (método <code class="prettyprint">bind</code>, conforme visto no segundo texto).</p>

<p>Em um terminal execute o comando <code class="prettyprint">docker-compose up --build</code> para subir os bancos e em outro terminal entremos na imagem <em>Alpine</em>, que contém o <em>Python</em>:</p>
<pre><code class="language-python">
In [1]: import bombril

In [2]: import models

In [3]: db = models.db

In [4]: data = bombril.read_csv()

In [5]: db.bind(provider=&#x27;postgres&#x27;, user=&#x27;postgres&#x27;, password=&#x27;top123&#x27;, host=&#x27;postgres&#x27;, database=&#x27;github&#x27;)

In [6]: db.generate_mapping(create_tables=True)

</code></pre>
<p>Como vimos, tudo funcionando, que tal fazer algumas coisas erradas?</p>

<p>Começando pelo <em>driver</em>:</p>
<pre><code class="language-python">
In [10]: db.bind(provider=&#x27;postgress&#x27;, user=&#x27;postgres&#x27;, password=&#x27;top123&#x27;, host=&#x27;postgres&#x27;, database=&#x27;github&#x27;)

... muitas linhas depois

ModuleNotFoundError: No module named &#x27;pony.orm.dbproviders.postgress&#x27;

</code></pre>
<p>Errando o usuário:</p>
<pre><code class="language-python">
In [11]: db.bind(provider=&#x27;postgres&#x27;, user=&#x27;postgress&#x27;, password=&#x27;top123&#x27;, host=&#x27;postgres&#x27;, database=&#x27;github&#x27;)

... muitas linhas depois

OperationalError: FATAL:  password authentication failed for user &quot;postgress&quot;

</code></pre>
<p>Que por acaso gera o mesmo erro para uma senha errada:</p>
<pre><code class="language-python">
In [12]: db.bind(provider=&#x27;postgres&#x27;, user=&#x27;postgres&#x27;, password=&#x27;top123....&#x27;, host=&#x27;postgres&#x27;, database=&#x27;github&#x27;)

... muitas linhas depois

OperationalError: FATAL:  password authentication failed for user &quot;postgres&quot;

</code></pre>
<p>Errando o <em>host</em>:</p>
<pre><code class="language-python">
In [13]: db.bind(provider=&#x27;postgres&#x27;, user=&#x27;postgres&#x27;, password=&#x27;top123&#x27;, host=&#x27;postgress&#x27;, database=&#x27;github&#x27;)

... muitas linhas depois

OperationalError: could not translate host name &quot;postgress&quot; to address: Name does not resolve

</code></pre>
<p>Finalmente, errando o nome do banco:</p>
<pre><code class="language-python">
In [14]: db.bind(provider=&#x27;postgres&#x27;, user=&#x27;postgres&#x27;, password=&#x27;top123&#x27;, host=&#x27;postgres&#x27;, database=&#x27;github..&#x27;)

... muitas linhas depois

</code></pre>
<p>OperationalError: FATAL:  database &quot;github..&quot; does not exist</p>

<p>O que nos leva a poder programar algo assim:</p>
<pre><code class="language-python">
In [15]: from pony import orm

In [16]: try:
    ...:     db.bind(provider=&#x27;postgres&#x27;, user=&#x27;postgres&#x27;, password=&#x27;top123&#x27;, host=&#x27;postgres&#x27;, database=&#x27;github..&#x27;)
    ...: except orm.OperationalError as e:
    ...:     print(e)
    ...:
FATAL:  database &quot;github..&quot; does not exist

</code></pre>
<p>Vamos arrumas as coisas:</p>
<pre><code class="language-python">
In [17]: db.bind(provider=&#x27;postgres&#x27;, user=&#x27;postgres&#x27;, password=&#x27;top123&#x27;, host=&#x27;postgres&#x27;, database=&#x27;github&#x27;)

In [18]: db.generate_mapping(create_tables=True)

</code></pre>
<p>Pegar um registro pela chave primária:</p>
<pre><code class="language-python">
In [19]: models.Github[1]
Out[19]: Github[1]

In [20]: models.Github[1].to_dict()
Out[20]: {&#x27;id&#x27;: 1, &#x27;key&#x27;: &#x27;current_user_url&#x27;, &#x27;value&#x27;: &#x27;https:&#x2F;&#x2F;api.github.com&#x2F;user&#x27;}

</code></pre>
<p>E tentar pegar um registro que não existe:</p>
<pre><code class="language-python">
In [21]: models.Github[99].to_dict()

</code></pre>
<p>... muitas linhas depois</p>
<pre><code class="language-python">
ObjectNotFound: Github[99]

</code></pre>
<p>Temos mais uma exceção para nossa coleção. Sempre lembrando, elas &quot;moram&quot; em  <code class="prettyprint">pony.orm</code>.</p>

<h4>Outras formas de pesquisa</h4>

<p>Uma forma de não gerar exceção e ficar no ifizinho é usar o <code class="prettyprint">get</code>:</p>
<pre><code class="language-python">
In [22]: response = models.Github.get(id=99)

In [23]: response

In [24]: type(response)
Out[24]: NoneType

In [25]: response is None
Out[25]: True

</code></pre>
<p>Porém, caso haja mais que uma resposta, teremos o belo estouro de uma exceção:</p>
<pre><code class="language-python">
In [26]: response = models.Github.get(lambda p: &#x27;a&#x27; in p.key)

... muitas linhas depois

MultipleObjectsFoundError: Multiple objects were found. Use select(...) to retrieve them

</code></pre>
<p>Interessante que o erro já vem com um conselho, então sigamo-lo:</p>
<pre><code class="language-python">
In [27]: response = models.Github.select(lambda p: &#x27;a&#x27; in p.key)

In [28]: response.count()
Out[28]: 17

</code></pre>
<p>Ou uma busca mais real:</p>
<pre><code class="language-python">
In [29]: response = models.Github.select(lambda p: &#x27;user&#x27; in p.key)

In [30]: response.show()
id|key                                 |value                                   
--+------------------------------------+----------------------------------------
1 |current_user_url                    |https:&#x2F;&#x2F;api.github.com&#x2F;user             
2 |current_user_authorizations_html_url|https:&#x2F;&#x2F;github.com&#x2F;settings&#x2F;connectio...
26|current_user_repositories_url       |https:&#x2F;&#x2F;api.github.com&#x2F;user&#x2F;repos{?ty...
29|user_url                            |https:&#x2F;&#x2F;api.github.com&#x2F;users&#x2F;{user}
30|user_organizations_url              |https:&#x2F;&#x2F;api.github.com&#x2F;user&#x2F;orgs
31|user_repositories_url               |https:&#x2F;&#x2F;api.github.com&#x2F;users&#x2F;{user}&#x2F;r...
32|user_search_url                     |https:&#x2F;&#x2F;api.github.com&#x2F;search&#x2F;users?q...

</code></pre>
<p>O <code class="prettyprint">show()</code> é muito útil em casos de <em>debug</em>, e poderíamos ter feito essa consulta em uma só linha com:</p>
<pre><code class="language-python">
models.Github.select(lambda p: &#x27;user&#x27; in p.key).show()

</code></pre>
<p>O <code class="prettyprint">select</code> retorna um iterável, então, pode-se transformá-lo em uma lista, ou, antes verificar pela contagem do próprio <code class="prettyprint">Pony</code>:</p>
<pre><code class="language-python">
In [31]: response = models.Github.select(lambda p: &#x27;non exist&#x27; in p.key)

In [32]: response.count()
Out[32]: 0

</code></pre>
<p>E para fechar nosso guia básico de erros, que tal tentar incluir um registro com chave primária existente? Aqui o experimento:</p>
<pre><code class="language-python">
In [33]: with orm.db_session:
...:     models.Github(id=1,key=&#x27;bla&#x27;, value=&#x27;ble&#x27;)

... muitas linhas depois

CacheIndexError: Cannot create Github: instance with primary key 1 already exists

</code></pre>
<p>Acredito que com essa enfadonha e nada exaustiva lista com situações de erro, eu possa ter pego e passado o espírito do tratamento de erros no <code class="prettyprint">Pony</code>. O importante é ter a documentação em mãos e um terminal com seu laboratório para testar as diversas situações e poder defender-se desse mundo hostil que é a comunicação em rede.</p>

<p>Nunca nos esqueçamos desses princípios:</p>
<pre><code class="language-python">
The Eight Fallacies of
Distributed Computing

Peter Deutsch

Essentially everyone,
when they first build a distributed application,
makes the following eight assumptions.

All prove to be false in the long run and all
cause big trouble and painful learning experiences.

1.  The network is reliable
2.  Latency is zero
3.  Bandwidth is infinite
4.  The network is secure
5.  Topology doesn&#x27;t change
6.  There is one administrator
7.  Transport cost is zero
8.  The network is homogeneous

</code></pre>
<blockquote>
<p><a href="https://www.researchgate.net/publication/322500050_Fallacies_of_Distributed_Computing_Explained">https://www.researchgate.net/publication/322500050<em>Fallacies</em>of<em>Distributed</em>Computing_Explained</a></p>
</blockquote>

<p>Então, sempre na defesa!</p>

    </div>
  </article>


    <footer class="footer">
        <div class="content has-text-centered">
            <p>
                O conteúdo e opiniões publicados nesse site são de inteira responsabilidade minha, e não refletem as opiniões e nem são endossados pelo meu empregador.
            </p>
                Este blog é gerido pelo <a target="_blank" href="https://github.com/paulohrpinheiro/gerablog">GeraBlog</a> e tem o layout dado pelo <a target="_blank" href="https://bulma.io">Bulma</a>.
            <p>
                Copyright © 2016-2022, <a target="_blank" href="https://twitter.com/PauloHRPinheiro">Paulo Henrique Rodrigues Pinheiro</a>.
            </p>
        </div>
    </footer>

</body>
</html>
